docs/user_authentication/third_party_token_management.md

```diff
@@ -0,0 +1,38 @@
+
+```sequence {theme: 'simple'}
+title Third Party IDP Token Management
+
+participant Pbench-Keycloak-server
+participant Pbench-Server
+participant Pbench-Dashboard
+participant Identity-Provider
+
+autonumber 1
+Pbench-Dashboard->Pbench-Server: GET login page
+Pbench-Server->Pbench-Dashboard: 200 login page
+Pbench-Dashboard->Identity-Provider: POST Authorization request
```

Comment:
I wonder about how this is framed. This isn't entirely accurate, nor, I think, is it what we really want to capture. In part, this is a matter of how/if we separate or characterize the distinction between "host system" and "Pbench Server"; but we've already talked about things like static artifact caching, including separating the Dashboard HTML/CSS/JS from the Server using NGINX or other means. Also, the browser basically loads the entire dashboard at startup (perhaps asynchronously), and "login page" is at best a bit imprecise because the dashboard is a single page app where the JS dynamically builds views by fiddling with the single HTML page DOM. There's no "login page" (as such), nor is that implementation detail of any relevance to the sequence diagram.

The dashboard will, in some manner, redirect to the IAM login page: whether that's google, Red Hat SSO, GitHub, Apple, or our own captive Keycloak. In some fashion that redirected URI will build a login/validation/registration/etc. page which we don't need to care about at all, and then redirect back to the dashboard with an authentication token it can save and present to the Pbench Server.

So maybe our sequence diagram starts with a page load to the IAM instance? In any case, this is between the browser and the IAM instance, with no Pbench Server involvement.

Another aspect to consider that might help focus this is that our authentication and "token management" workflow should be equally applicable to a CLI model where there's no browser: the CLI, running on some anonymous host, will use an IAM API to authenticate and acquire a token, which it will then pass to the server. No web page load at all, no dashboard. Maybe we'll need separate sequence diagrams: but I hope the concepts and steps are similar enough that we don't need to do that.

Reply:
It's more than that: the sequence diagram deals with the authentication workflow, not the entire Pbench/dashboard workflow. Loading the dashboard code *isn't* part of "authentication" at all. The dashboard is loaded and running in the browser; the user then initiates the workflow by clicking on an authentication link, which initiates a redirect to the identity manager. (And for a CLI there's no user clicking on a UI element, or any UI element at all, so I'd prefer to just skip that step.) Maybe we start with the client getting the IAM identification (e.g., from an "authentication" section of the endpoints API response) to establish context before reaching out to the identity provider... but in any case, loading dashboard code has nothing to do with this flow.

Reply:
I didn't spend a ton of time on this, but here's a possible change to (I think) improve the flow a bit and remove a few bugs. (E.g., you list 401 as the alternative to success at the end; but as we've gone to all the trouble of getting a token from the IAM, we're really talking about `403`/FORBIDDEN here, not `401`/UNAUTHORIZED.)

```
title Third Party IDP Token Management

participant Pbench-Keycloak-server
participant Pbench-Server
participant "Authentication-client" as Client
participant Identity-Provider

autonumber 1
Client->Pbench-Server:GET IAM scheme & link
Pbench-Server->Client:200 return authentication schema/link
Client->Identity-Provider: POST Authorization request
note over Identity-Provider: Authenticate User and request \nuser consent for Pbench
Identity-Provider->Client: 200 Authorization code response
Client->Identity-Provider: POST authorization code (Token request)
Identity-Provider->Client: 200 Access/Refresh Token payload response 
note over Identity-Provider,Client: {"access_token":"<access_token>",\n"expires_in": 300,\n"refresh_expires_in": 1800,\n"refresh_token": "<refresh_token>",\n"token_type": "Bearer",\n"id_token": "<id_token>",...}
// Here the dashboard sends the newly acquired access token to the Pbench Server to start the token exchange process.
Client->Pbench-Server: POST /api/v1/pbench-token request (Bearer: Access Token 1)
Pbench-Server->Identity-Provider: GET public Key
Identity-Provider->Pbench-Server: 200 Public key
Pbench-Server->Pbench-Server: Validate the access token
Pbench-Server->Identity-Provider: [Optional] POST userinfo endpoint
Identity-Provider->Pbench-Server:  200 userinfo response
note over Pbench-Server,Client: {"family_name":"<surname>",\n"sub": <user_id>,\n"email_verified": <bool>,\n"email": "<email>",\n"given_name": "given_name",...}
// We exchange the Third-party IDP token with our internal Pbench token that should be used for accessing Pbench server resources.
// Pbench token might contain more metadata about user such as role/group associations.
Pbench-Server->Pbench-Keycloak-server: Exchange IDP access token \n(provided by the Pbench dashboard) \nwith Pbench token
Pbench-Keycloak-server->Pbench-Server: 200 new Pbench access_token and refresh_token
// Finally, the Pbench Server responds to the dashboard with the new pbench token in the response payload.
Pbench-Server->Client:200 /api/v1/pbench-token reponse (new 'Pbench Token' in payload)
Client->Pbench-Server:POST /api/v1/<API> request (Bearer: Pbench Token)
note over Pbench-Server: Validation and Role extraction \nfrom the Pbench token
alt Access allowed
Pbench-Server->Client: 200 /api/v1/<restricted_endpoint> reponse
else Access Forbidden
Pbench-Server->Client:403 /api/v1/<restricted_endpoint> reponse
end
```

Reply:
So the `authentication_client` can be Pbench dashboard or some other client like CLI users. I like this generalization, there might be certain differences in the flow for the CLI users like we would give them a redirect link instead of redirecting them to the identity provider that a user has to copy and go to the browser login and come back and the Pbench server would have their tokens ready.

Reply:
I think Dave's improvements are good (although, I'm starting to see a bunch of typos that I didn't see before...I dunno whether they are all Dave's or whether some are Nikhil's 😀).

However, I suspect that the `authentication_client` is always going to be a browser-based GUI; otherwise, I think the dance is going to be too cumbersome.  (E.g., note that Janos's demo today involved going to a browser to get the token....)

In actuality, for the CLI, I think we're going to end up following the lead of AWS, COPR, Quay.io, et al. and use an "access key" (etc.) which is basically a permanent token that you get from a GUI -- the CLI itself won't accept credentials.

This also allows us to remove the identity-provider from the CLI flow, in general, which will be useful, given that the typical CLI uses are in scripts which typically aren't interactive.